clients/venice-client/src/main/java/com/linkedin/venice/fastclient/InternalAvroStoreClient.java

```diff
@@ -1,58 +1,197 @@
+    Map<K, CompletableFuture<V>> valueFutures = new HashMap<>();
+    keys.forEach(k -> {
+      valueFutures.put(k, (get(new GetRequestContext(true), k)));
+    });
+    CompletableFuture.allOf(valueFutures.values().toArray(new CompletableFuture[0]))
+        .whenComplete(((aVoid, throwable) -> {
+          if (throwable != null) {
+            resultFuture.completeExceptionally(throwable);
+          }
+          Map<K, V> resultMap = new HashMap<>();
+          valueFutures.forEach((k, f) -> {
+            try {
+              resultMap.put(k, f.get());
+            } catch (Exception e) {
+              resultFuture.completeExceptionally(
+                  new VeniceClientException("Failed to complete future for key: " + k.toString(), e));
+            }
+          });
+          resultFuture.complete(resultMap);
+        }));
+
+    return resultFuture;
   }
 
-  protected abstract CompletableFuture<Map<K, V>> batchGet(BatchGetRequestContext<K, V> requestContext, Set<K> keys)
-      throws VeniceClientException;
+  @Override
+  public final void streamingBatchGet(Set<K> keys, StreamingCallback<K, V> callback) throws VeniceClientException {
+    streamingBatchGet(new BatchGetRequestContext<>(true), keys, callback);
+  }
 
-  public void streamingBatchGet(Set<K> keys, StreamingCallback<K, V> callback) throws VeniceClientException {
-    streamingBatchGet(new BatchGetRequestContext<K, V>(), keys, callback);
+  @Override
+  public final CompletableFuture<VeniceResponseMap<K, V>> streamingBatchGet(Set<K> keys) throws VeniceClientException {
+    return streamingBatchGet(new BatchGetRequestContext<>(true), keys);
   }
 
-  public CompletableFuture<VeniceResponseMap<K, V>> streamingBatchGet(Set<K> keys) throws VeniceClientException {
-    return streamingBatchGet(new BatchGetRequestContext<K, V>(), keys);
+  protected final CompletableFuture<VeniceResponseMap<K, V>> streamingBatchGet(
+      BatchGetRequestContext<K, V> requestContext,
+      Set<K> keys) {
+    int keySize = keys.size();
+    // keys that do not exist in the storage nodes
+    Queue<K> nonExistingKeys = new ConcurrentLinkedQueue<>();
+    VeniceConcurrentHashMap<K, V> valueMap = new VeniceConcurrentHashMap<>();
+    CompletableFuture<VeniceResponseMap<K, V>> streamingResponseFuture = new VeniceResponseCompletableFuture<>(
+        () -> new VeniceResponseMapImpl<>(valueMap, nonExistingKeys, false),
+        keySize,
+        Optional.empty());
+    streamingBatchGet(requestContext, keys, new StreamingCallback<K, V>() {
+      @Override
+      public void onRecordReceived(K key, V value) {
+        if (value == null) {
+          nonExistingKeys.add(key);
+        } else {
+          valueMap.put(key, value);
+        }
+      }
+
+      @Override
+      public void onCompletion(Optional<Exception> exception) {
+        requestContext.complete();
+        if (exception.isPresent()) {
+          streamingResponseFuture.completeExceptionally(exception.get());
```

Comment:
With this implementation, this function doesn't return partial response if some of the routes throws an exception because of a backend error.
And `streamingResponseFuture.get()` will propagate the exception to the caller.

We may need to tweak `VeniceResponseCompletableFuture#get` function to catch all exceptions and always return partial response.

Reply:
Yeah, looks like the current test cases are written to expect an exception: https://github.com/linkedin/venice/blob/main/clients/venice-client/src/test/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClientTest.java#L732

I think you're right that the exception should not be thrown in `get()` as well. I'll override `VeniceResponseCompletableFuture#get` to do it this way - and not throw exceptions.